lib/identity/auth.rb

```diff
@@ -226,80 +223,22 @@ class Auth < Sinatra::Base
-      end
-
-      # clear anything that might be left over in the session
-      @cookie.authorize_params = nil
-
-      begin
-        # Have the user login if:
-        # - We have no session for them
-        # - The client requested that they login
-        if !@cookie.access_token || params[:prompt] == 'login'
-          raise Identity::Errors::LoginRequired
-        end
-
-        # Try to perform an access token refresh if we know it's expired. At
-        # the time of this writing, refresh tokens last 30 days (much longer
-        # than the short-lived 2 hour access tokens).
-        if Time.now > @cookie.access_token_expires_at
-          perform_oauth_refresh_dance
-        end
-
-        # redirects back to the oauth client on success
-        authorize(authorize_params, params[:authorize] == "Allow")
-      # given client_id wasn't found (API throws a 400 status)
-      rescue Excon::Errors::BadRequest
-        flash[:error] = "Unknown OAuth client."
-        redirect to("/login")
-      # we couldn't track the user's session meaning that it's likely been
-      # destroyed or expired, redirect to login
-      rescue Excon::Errors::NotFound
-        redirect to("/login")
-      # user needs to login.
-      rescue Identity::Errors::LoginRequired
-        flash[:link_account] = true
-        @cookie.post_signup_url = request.url
-        @cookie.authorize_params = authorize_params
-        redirect to("/login")
-      # refresh token dance was unsuccessful
-      rescue Excon::Errors::Unauthorized
-        @cookie.authorize_params = authorize_params
-        redirect to("/login")
-      rescue Identity::Errors::PasswordExpired => e
-        flash[:error] = e.message
-        redirect to("/account/password/reset")
-      rescue Identity::Errors::SuspendedAccount => e
-        flash[:error] = e.message
-        redirect to("/login")
-      # client not yet authorized; show the user a confirmation dialog
-      rescue Identity::Errors::UnauthorizedClient => e
-        @cookie.authorize_params = authorize_params
-        @client = e.client
-        @scope  = @cookie && @cookie.authorize_params["scope"] || nil
-        @deny_url = build_uri(@client["redirect_uri"], { error: "access_denied" })
-        slim :"clients/authorize", layout: :"layouts/purple"
-      # for example, "invalid scope"
-      rescue Excon::Errors::UnprocessableEntity => e
-        flash[:error] = decode_error(e.response.body)
+    def redirect_or_authorize
+      if @cookie.sso_entity && Config.sso_base_url
+        # If the user is using SSO, have them auth with the SSO service
+        @cookie.authorize_params = get_authorize_params
+        redirect to("#{Config.sso_base_url}/#{@cookie.sso_entity}")
+      elsif !@cookie.access_token || params[:prompt] == "login"
+        # Have users login if they don't have a session, or the client
+        # has requested an explicit login
+        @cookie.authorize_params = get_authorize_params
         redirect to("/login")
```

Comment:
I have no idea if they are safe to omit. Good catch, because this was unintentional!